### PR TITLE
Add example for IAM authentication

### LINKED ISSUE #75: [FEATURE] Add IAM authentication option
If people want to connect to an Amazon OpenSearch Service cluster, they have to figure out how to use this client alongside some third-party signing library and/or the AWS SDK, which is non-trivial due to how we (by design) conceal the underlying HTTP requests that the client makes. We should offer IAM signing as an option, just like the OpenSearch CLI does.

**Is your feature request related to a problem?**
If you want to use this client with an Amazon OpenSearch Service cluster that has IAM authentication rather than basic authentication, good luck.

**What solution would you like?**

When initializing the client, an AuthType option. If basic (or null), accept username and password. If IAM, accept options for access key, secret key, session token, region, and service.

**What alternatives have you considered?**

An additional Go IAM signing library for use on top of the client. But given that it would only work with the client, it seems better and easier to just add it to the client.

**Do you have any additional context?**

https://github.com/opensearch-project/OpenSearch/issues/1400

### PR DESCRIPTION
**Pull request by rmasclef**
### Description
Add an example explaining how to perform authentication using IAMs. 
 
### Issues Resolved
#75 
 
### Check List
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: USER_GUIDE.md [hunk 1]
```diff
@@ -124,4 +124,72 @@ func main() {
...
+	"io"
+	"log"
+
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/opensearch-project/opensearch-go"
+	"github.com/opensearch-project/opensearch-go/opensearchapi"
+	requestsigner "github.com/opensearch-project/opensearch-go/signer/aws"
+)
+
+const domainEndpoint = "https://vpc-test-opensearch-with-iams-lpzna4633opqnm.eu-west-1.es.amazonaws.com/"
```
Comment (VijayanB) [inline USER_GUIDE.md:153]: const endpoint = "" // e.g. https://opensearch-domain.region.com

Comment (rmasclef) [reply USER_GUIDE.md]: Done 😉

File: USER_GUIDE.md [hunk 2]
```diff
@@ -124,4 +124,72 @@ func main() {
     fmt.Println("deleting index", deleteIndexResponse)
 }
 
-```
\ No newline at end of file
+```
+
+## How to use IAMs as authentication method
+
+Before starting, we strongly recommend to read the full AWS documentation regarding IAMs for Opensearch.
```
Comment (keithhc2) [inline USER_GUIDE.md:131]: "Before starting, we strongly recommend reading the full AWS documentation regarding using IAM credentials to sign to requests to OpenSearch APIs."

File: USER_GUIDE.md [hunk 3]
```diff
@@ -124,4 +124,72 @@ func main() {
...
+See https://docs.aws.amazon.com/opensearch-service/latest/developerguide/ac.html
```
Comment (keithhc2) [inline USER_GUIDE.md:132]: I would recommend linking off to the developer guide with some text, not just a hyperlink.

See [Identity and Access Management in Amazon OpenSearch Service](https://docs.aws.amazon.com/opensearch-service/latest/developerguide/ac.html).

File: USER_GUIDE.md [hunk 4]
```diff
@@ -124,4 +124,72 @@ func main() {
...
+
+> Even if you configure a completely open resource-based access policy, all requests to the OpenSearch Service configuration API must be signed. If your policies specify IAM users or roles, requests to the OpenSearch APIs also must be signed using AWS Signature Version 4.
+https://docs.aws.amazon.com/opensearch-service/latest/developerguide/ac.html#managedomains-signing-service-requests
+
+Here is an example of Go code showing how to sign each opensearch request and automatically search for AWS credentials from ~/aws folder or env vars:
```
Comment (keithhc2) [inline USER_GUIDE.md:137]: Capitalize OpenSearch?

**Review by keithhc2 — DISMISSED** [review]

File: USER_GUIDE.md [hunk 5]
```diff
@@ -124,4 +124,72 @@ func main() {
...
+	ping := opensearchapi.PingRequest{}
+
+	resp, err := ping.Do(ctx, client)
+	if err != nil {
+		log.Fatal(err)
+	}
+	defer resp.Body.Close()
+
+	if resp.IsError() {
+		log.Println("ping resp code", resp.StatusCode)
```
Comment (VijayanB) [inline USER_GUIDE.md:183]:
```suggestion
		log.Println("ping response status ", resp.Status())
```

File: USER_GUIDE.md [hunk 6]
```diff
@@ -124,4 +124,72 @@ func main() {
...
+
+## How to use IAMs as authentication method
+
+Before starting, we strongly recommend reading the full AWS documentation regarding using IAM credentials to sign requests to OpenSearch APIs.
+See [Identity and Access Management in Amazon OpenSearch Service.](https://docs.aws.amazon.com/opensearch-service/latest/developerguide/ac.html)
+
+> Even if you configure a completely open resource-based access policy, all requests to the OpenSearch Service configuration API must be signed. If your policies specify IAM users or roles, requests to the OpenSearch APIs also must be signed using AWS Signature Version 4.
+See [Managed Domains signing-service requests.](https://docs.aws.amazon.com/opensearch-service/latest/developerguide/ac.html#managedomains-signing-service-requests)
+
+Here is some sample Go code that shows how to sign each OpenSearch request and automatically search for AWS credentials from the ~/aws folder or env vars:
```
Comment (VijayanB) [inline USER_GUIDE.md:137]:
```suggestion
Here is some sample Go code that shows how to sign each OpenSearch request and automatically search for AWS credentials from the ~/.aws folder or environment variables:
```

**Comment by VijayanB** [comment]
Looks good. Just minor comment. Can you exclude example endpoint from check here: https://github.com/opensearch-project/opensearch-go/blob/main/.github/workflows/links.yml#L21 to fix Link Checker failure. Thanks

**Comment by rmasclef** [comment]
> Looks good. Just minor comment. Can you exclude example endpoint from check here: https://github.com/opensearch-project/opensearch-go/blob/main/.github/workflows/links.yml#L21 to fix Link Checker failure. Thanks

https://github.com/opensearch-project/opensearch-go/pull/93/files#diff-837695f5e5bb7126693b21d45c2ad50d412f36496f0bd88aa88b7b3d874eeae8R153

I'm not sure I understand what you want me to do.

Add `https://opensearch-domain.region.com` to the excluded links ?

**Comment by rmasclef** [comment]
> NB : I found a little bit annoying to wait for approval to see if the link checker passes or not 😬

**Review by dblock — APPROVED** [review]

**Comment by dblock** [comment]
@VijayanB I kicked CI/CD again for the failed jobs and it passed, CR/merge?

**Review by VijayanB — APPROVED** [review]

**Review by dblock — APPROVED** [review]

**Pull request merged**